src/coprocessor/dag/builder.rs

```diff
@@ -38,20 +40,83 @@ pub struct DAGBuilder;
-        false
+    pub fn can_build_batch(exec_descriptors: &[executor::Executor]) -> bool {
+        use cop_datatype::EvalType;
+        use cop_datatype::FieldTypeAccessor;
+        use std::convert::TryFrom;
+
+        for ed in exec_descriptors {
+            match ed.get_tp() {
+                ExecType::TypeTableScan => {
+                    let descriptor = ed.get_tbl_scan();
+                    for column in descriptor.get_columns() {
+                        let eval_type = EvalType::try_from(column.tp());
+                        if eval_type.is_err() {
+                            debug!(
+                                "Coprocessor request cannot be batched";
+                                "unsupported_column_tp" => ?column.tp(),
+                            );
+                            return false;
+                        }
+                    }
+                }
+                _ => {
+                    debug!(
+                        "Coprocessor request cannot be batched";
+                        "unsupported_executor_tp" => ?ed.get_tp(),
+                    );
+                    return false;
+                }
+            }
+        }
+        true
     }
 
     // Note: `S` is `'static` because we have trait objects `Executor`.
     pub fn build_batch<S: Store + 'static>(
-        _executor_descriptors: Vec<executor::Executor>,
-        _store: S,
-        _ranges: Vec<KeyRange>,
-        _config: Arc<EvalConfig>,
+        executor_descriptors: Vec<executor::Executor>,
+        store: S,
+        ranges: Vec<KeyRange>,
+        config: Arc<EvalConfig>,
     ) -> Result<Box<dyn BatchExecutor>> {
-        // Currently no batch executors and this function should never be called, so unreachable.
-        unreachable!()
+        let mut executor_descriptors = executor_descriptors.into_iter();
+        let mut first_ed = executor_descriptors
+            .next()
+            .ok_or_else(|| Error::Other(box_err!("No executors")))?;
+
+        let mut executor: Box<dyn BatchExecutor>;
+
+        match first_ed.get_tp() {
+            ExecType::TypeTableScan => {
+                // TODO: Use static metrics.
+                COPR_EXECUTOR_COUNT.with_label_values(&["tblscan"]).inc();
+
+                let mut descriptor = first_ed.take_tbl_scan();
+                let columns_info = descriptor.take_columns().into_vec();
+                executor = Box::new(BatchTableScanExecutor::new(
+                    store,
+                    config.clone(),
+                    columns_info,
+                    ranges,
+                    descriptor.get_desc(),
+                )?);
+            }
+            _ => {
+                return Err(Error::Other(box_err!(
+                    "Unexpected first executor {:?}",
+                    first_ed.get_tp()
+                )));
+            }
+        }
+
+        if executor_descriptors.next().is_some() {
```

Comment:
Can you explain more about this?

Reply:
Means there are more than one executors in the protobuf file.

Reply:
Can you move it before Line 87 or move it out of the function, only provide one executor as the parameter?

Reply:
Let's don't do this, because we *should* accept multiple executors for this function. It accepts one executor for now only because only table scan is implemented.

I added a comment in the code to explain why we need to check the executor count for now. Do you think it's acceptable?